seastar-macros/src/lib.rs

```diff
@@ -42,3 +42,70 @@ pub fn test(
 
     output.into()
 }
+
+#[proc_macro_attribute]
+#[cfg(not(test))] // Work around for rust-lang/rust#62127
+pub fn main(
+    args: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let args = syn::parse_macro_input!(args as proc_macro2::TokenStream);
+    let input = syn::parse_macro_input!(item as syn::ItemFn);
+
+    let output = &input.sig.output;
+    let inputs = &input.sig.inputs;
+    let name = &input.sig.ident;
+    let body = &input.block;
+    let attrs = &input.attrs;
+
+    if input.sig.asyncness.is_none() {
+        let msg = "the async keyword is missing from the function declaration";
+        return syn::Error::new_spanned(input.sig, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    if name != "main" {
+        let msg = "only the main function is allowed to use #[seastar::main]";
+        return syn::Error::new_spanned(input.sig.ident, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    if !inputs.is_empty() {
+        let msg = "the main function cannot accept arguments";
+        return syn::Error::new_spanned(input.sig, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    if !args.is_empty() {
+        let msg = "arguments for #[seastar::test] are not supported yet";
+        return syn::Error::new_spanned(args, msg).to_compile_error().into();
+    }
+
+    let ret_type = match output {
+        syn::ReturnType::Type(_, ty) => ty.clone(),
+        syn::ReturnType::Default => syn::parse_quote! { () },
+    };
+
+    let output = quote::quote! {
+        #(#attrs)*
+        fn main() #output {
+            let ret_holder : std::rc::Rc<std::cell::Cell<Option<#ret_type>>> = Default::default();
+
+            let ret_holder_clone = ret_holder.clone();
+            let fut = async move {
+                let ret = #body;
+                ret_holder_clone.set(Some(ret));
+                Ok(())
+            };
+
+            let mut app = seastar::AppTemplate::default();
+            app.run_void(std::env::args(), fut);
+            ret_holder.take().unwrap()
```

Comment:
I think `Rc<Cell<Option<T>>>` is overkill here - oneshot channel would seem more appropriate.

Reply:
Now that I think of it, it's not necessarily true, current solution if pretty low overhead, channel may have some synchronization which would be more heavyweight. If there is some crate that has non-Send one-shot channel with comparable overhead to current solution, we could think about using it - if not, then let's stick with current solution.

Reply:
Here's a fast one I found:
https://crates.io/crates/async-oneshot

Reply:
I benched `Rc<Cell<Option<T>` against `async_oneshot::oneshot::<T>` and here are the results:
### `Rc<Cell<Option<T>`
create_destroy          time:   [22.846 ns 22.879 ns 22.942 ns]        
send/success            time:   [39.096 ns 40.330 ns 41.581 ns]        
try_recv/success        time:   [38.105 ns 39.078 ns 40.054 ns]                             

### `async_oneshot::oneshot::<T>`
create_destroy          time:   [44.632 ns 44.688 ns 44.748 ns]                            
send/success            time:   [9.0997 ns 9.2451 ns 9.3948 ns]            
try_recv/success        time:   [41.189 ns 42.524 ns 43.838 ns]

Reply:
Ok, let's stick with `Rc<Cell<Option<T>>>` - it's inelegant, but I wouldn't want to sacrifice performance.